yamlpatch/patch.go

```diff
@@ -0,0 +1,264 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package yamlpatch is a library for applying RFC6902 JSON patches to yaml documents.
+// It leverages go-yaml v3's Node type to preserve comments, ordering, and most formatting.
+package yamlpatch
+
+import (
+	"bytes"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	// defaultIndentSpaces configures the yaml encoder used to marshal patched objects.
+	defaultIndentSpaces = 2
+)
+
+// Apply the patch to a yaml document provided in originalBytes and return the updated content.
+// A best effort is made to minimize changes outside the patched paths but some whitespace changes are unavoidable.
+func Apply(originalBytes []byte, patch Patch) ([]byte, error) {
+	return ApplyIndent(originalBytes, patch, defaultIndentSpaces)
+}
+
+// ApplyIndent the patch to a yaml document provided in originalBytes and return the updated content.
+// indentSpaces configures the indentation used to marshal patched objects.
+// A best effort is made to minimize changes outside the patched paths but some whitespace changes are unavoidable.
+func ApplyIndent(originalBytes []byte, patch Patch, indentSpaces int) ([]byte, error) {
```

Comment:
Thought about this a bit more, and propose the following path forward:

* As part of this initial PR, let's just remove `ApplyIndent` and only have the `Apply` function. This means that indentation level can't be configured as part of the initial PR, but think that should be fine.
* I will do this as a follow-up PR, but I propose that we follow the general pattern of things like `json.Marshal` here and effectively have a `PatchApplier` struct that can be created and configured (whether it be via params or functions) to apply patches. The applier can be configured to do things like set indentation level, control whether values are marshaled as JSON or YAML, etc.
* Implementation of `ApplyIndent` will be updated to effectively use the default `PatchApplier` struct/settings